### ISSUE TITLE
ProxyBuilder.getMethodsToProxy() attempts to proxy final method on child class if non-final on superclass

### ISSUE THREAD

**Issue by aviverette** [linkedin/dexmaker#31]
Found in revision: f54f6eac0c81120f53265b30adf9ce602e8dfc41

Encountered while attempting to use Mockito's spy() method to create a real partial mock of a FrameLayout:

```
FrameLayout parent = spy(new FrameLayout(mActivity));
```

ProxyBuilder recursively searches for methods starting from the root class, which results in non-final superclass methods being added to the methods-to-override sink. When the search reaches a child class where the method is marked final, it is added to the set of seen final methods; however, it is not removed from the set of methods-to-override.

**Comment by stjj89** [comment]
I ran into a similar, related issue when attempting to Mock android.widget.Spinner in a unittest, which led to an error indicating that the generated Proxy contained a getParent() method, which overrides the final method in the parent class android.view.View:

java.lang.LinkageError: Method android.view.ViewParent Spinner_Proxy.getParent() overrides final method in class Landroid/view/View; (declaration of 'Spinner_Proxy' 
appears in /data/user_de/0/com.android.settings/cache/Generated_1011145298.jar)

In my case, a proxy was generated for a final method inherited from a parent class.

I found the issue in dexmaker's ProxyBuilder.getMethodsToProxyRecursive(), which is called from ProxyBuilder.buildProxyClass() to get a set of methods to create proxies for. This recursive method basically goes up the child-parent class hierarchy, adding each method we encounter to the set of methods to proxy, as long as the method is not final or static, and is not named "finalize". However, after finishing examining all methods for a given class and before moving up to the next level of the hierarchy, it traverses all interface classes that the current class implements, and scan for methods to proxy in those interfaces. The problem is that these interfaces might contain non-final methods that we will later encounter as final methods in parent classes. Here's how we run into this issue with Spinner:

Spinner -> AbsSpinner (parent of Spinner) -> AdapterView (parent of AbsSpinner) -> ViewGroup (parent of AdapterView) -> ViewParent (interface implemented by ViewGroup)

At this point in ViewParent, we find getParent() and add it to the set, since it is not final. Then, we continue the traversal:

-> ViewManager (interface implemented by ViewGroup) -> View (parent of ViewGroup)

Now, in View, we see getParent() again, which is final. We do not add this method to our set, but we do not do anything about the existing entry that is already in the set.

Finally, when the Spinner_Proxy is generated, it contains a getParent() proxy method, which overrides the final getParent() method in the parent class View, which leads to a runtime error.

A fix for this will be to have dexmaker delete any existing methods in the set of methods to proxy whenever we find a final method in ProxyBuilder.getMethodsToProxy(). Patch incoming.
